### PR TITLE
Receive streams in WebRTC (and MediaStreamTrack support)

### PR DESCRIPTION
**Pull request by Manishearth**
This adds the `ontrack` event handler to webrtc, and all the `MediaStreamTrack` stuff necessary to make it work.

WebRTC has the ability to group media tracks into streams using MSIDs, but I haven't yet figured out how to do this. For now, `ontrack` should work.

This _should_ be complete, but it hasn't yet been tested (hence the WIP)

r? @ferjm or @jdm

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23342)

### CODE REVIEW

File: components/script/dom/rtctrackevent.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            Atom::from(type_),
+            init.parent.bubbles,
+            init.parent.cancelable,
+            &init.track,
+        ))
+    }
+}
+
+impl RTCTrackEventMethods for RTCTrackEvent {
+    // https://html.spec.whatwg.org/multipage/#dom-RTCTrackEvent-track
```
Comment (jdm) [inline components/script/dom/rtctrackevent.rs]: Unlikely URL.

**Review by jdm — APPROVED** [review]
Nice!

**Comment by Manishearth** [comment]
omg it works.


image

**Comment by Manishearth** [comment]
Testing code at https://github.com/Manishearth/gstwebrtc-demos/tree/servo

**Comment by Manishearth** [comment]
@ferjm thanks for all the work in https://github.com/servo/servo/pull/23157, it Just Works™ here

File: components/script/dom/mediastream.rs [hunk 2]
```diff
@@ -83,4 +83,19 @@ impl MediaStreamMethods for MediaStream {
...
+    fn AddTrack(&self, track: &MediaStreamTrack) {
+        let existing = self.tracks.borrow().iter().find(|x| *x == &track).is_some();
+
+        if existing {
+            return;
+        }
+        self.add_track(track)
+    }
+
+    /// https://w3c.github.io/mediacapture-main/#dom-mediastream-addtrack
```
Comment (ferjm) [inline components/script/dom/mediastream.rs]: This should be `https://w3c.github.io/mediacapture-main/#dom-mediastream-removetrack`

File: components/script/dom/mediastream.rs [hunk 3]
```diff
@@ -98,4 +98,13 @@ impl MediaStreamMethods for MediaStream {
     fn RemoveTrack(&self, track: &MediaStreamTrack) {
         self.tracks.borrow_mut().retain(|x| *x != track);
     }
+
+    /// https://w3c.github.io/mediacapture-main/#dom-mediastream-clone
+    fn Clone(&self) -> DomRoot<MediaStream> {
+        let new = MediaStream::new(&self.global());
+        for track in &*self.tracks.borrow_mut() {
```
Comment (ferjm) [inline components/script/dom/mediastream.rs]: I think this does not need to be a mutable borrow.

**Review by ferjm — APPROVED** [review]
r=me with the comments addressed. Thanks!

**Comment by CYBAI** [comment]
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: If this is failing: DANGER, are you sure you want to expose the new interface RTCTrackEvent to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false", 
    "stack": "test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:70:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1587:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\ntest_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3\n@http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1\n", 
    "subtest": "Interfaces exposed on the window", 
    "test": "/_mozilla/mozilla/interfaces.html", 
    "line": 310798, 
    "action": "test_result", 
    "expected": "PASS"
}
```

**Comment by Manishearth** [comment]
Forgot to pref-gate

**Pull request merged**